### PR TITLE
[Merged by Bors] - feat: improvements to `congr!` and `convert`

### PR DESCRIPTION
**Pull request by kmill**
* There is now configuration for `congr!`, `convert`, and `convert_to` to control parts of the congruence algorithm, in particular transparency settings when applying congruence lemmas.
* `congr!` now applies congruence lemmas with reducible transparency by default. This prevents it from unfolding definitions when applying congruence lemmas. It also now tries both the LHS-biased and RHS-biased simp congruence lemmas, with a configuration option to set which it should try first.
* There is now a new `HEq` congruence lemma generator that gives each hypothesis access to the proofs of previous hypotheses. This means that if you have an equality `⊢ ⟨a, x⟩ = ⟨b, y⟩` of sigma types, `congr!` turns this into goals `⊢ a = b` and `⊢ a = b → HEq x y` (note that `congr!` will also auto-introduce `a = b` for you in the second goal). This congruence lemma generator applies to more cases than the simp congruence lemma generator does.
* `congr!` (and hence `convert`) are more careful about applying lemmas that don't force definitions to unfold. There were a number of cases in mathlib where the implementation of `congr` was being abused to unfold definitions.
* With `set_option trace.congr! true` you can see what `congr!` sees when it is deciding on congruence lemmas.
* There is also a bug fix in `convert_to` to do `using 1` when there is no `using` clause, to match its documentation.

Note that `congr!` is more capable than `congr` at finding a way to equate left-hand sides and right-hand sides, so you will frequently need to limit its depth with a `using` clause. However, there is also a new heuristic to prevent considering unlikely-to-be-provable type equalities (controlled by the `typeEqs` option), which can help limit the depth automatically.

There is also a predefined configuration that you can invoke with, for example, `convert (config := .unfoldSameFun) h`, that causes it to behave more like `congr`, including using default transparency when unfolding.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: test/convert.lean [hunk 1]
```diff
@@ -11,52 +12,60 @@ example (α β : Type) (h : ∀ α β : Type, α = β) (b : β) : α := by
...
 section convert_to
 
 example {α} [AddCommMonoid α] {a b c d : α} (H : a = c) (H' : b = d) : a + b = d + c := by
-  convert_to c + d = _
+  convert_to c + d = _ using 2
   rw [add_comm]
 
 example {α} [AddCommMonoid α] {a b c d : α} (H : a = c) (H' : b = d) : a + b = d + c := by
-  convert_to c + d = _ using 2
+  convert_to c + d = _ using 1
```
Comment (fpvandoorn) [inline test/convert.lean:48]:
```suggestion
  convert_to c + d = _
```
?

File: Mathlib/Tactic/Congr!.lean [hunk 2]
```diff
@@ -20,38 +20,67 @@ See the syntax docstring for more details.
...
+  /-- For passes that synthesize a congruence lemma using one side of the equality,
+  we run the pass both for the left-hand side and the right-hand side. If `preferLHS` is `true`
+  then we start with the left-hand side.
+
+  This can be used to control which side's definitions are expanded when applying the
+  congruence lemma (if `preferLHS = true` then the RHS can be expanded). -/
+  preferLHS : Bool := true
+  /-- Allow one or both sides to be a partial application.
+  When false, given an equality `f a b = g x y` this means we never consider
+  proving `f a = g x`. -/
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:44]: Do we still consider proving `f = g`?

File: Mathlib/Tactic/Congr!.lean [hunk 3]
```diff
@@ -20,38 +20,67 @@ See the syntax docstring for more details.
...
+  partialApp : Bool := true
+  /-- Whether to require both sides in an application to have defeq functions.
+  That is, if true, `f a = g x` is only considered if `f` and `g` are defeq. -/
+  sameFun : Bool := false
+  /-- The maximum number of arguments to consider at a time when applying synthesized congruence
+  lemmas between function applications. `none` means no limit.
+  With non-dependent functions, `some 1` causes `congr!` lets you control how many arguments
+  it ends up processing using the iteration limit. -/
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:52]: Please reformulate (the sentence is not grammatically correct and I don't understand it).

File: Mathlib/Tactic/Congr!.lean [hunk 4]
```diff
@@ -20,38 +20,67 @@ See the syntax docstring for more details.
...
+  maxArgs : Option Nat := none
+  /-- Whether to consider equalities between types that are don't look provable according to
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:54]:
```suggestion
  /-- Whether to consider equalities between types that don't look provable according to
```

File: Mathlib/Tactic/Congr!.lean [hunk 5]
```diff
@@ -20,38 +20,67 @@ See the syntax docstring for more details.
...
+  a heuristic. When `false`, then causes the main congruence generator to apply a "possibly
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:55]:
```suggestion
  a heuristic. When `false`, that causes the main congruence generator to apply a "possibly
```

File: Mathlib/Tactic/Congr!.lean [hunk 6]
```diff
@@ -20,38 +20,67 @@ See the syntax docstring for more details.
...
+  provable" heuristic when there would be a type equality." Note: we always allow equalities of
```
Comment (fpvandoorn) [inline Mathlib/Tactic/Congr!.lean:56]:
```suggestion
  provable" heuristic when there would be a type equality." 
  When `true` we consider all type equalities. Note: we always allow equalities of
```
?

**Review by fpvandoorn — COMMENTED** [review]
LGTM

bors d+

File: test/convert.lean (hunk 1 above)
Comment (kmill) [reply test/convert.lean:48]: I guess this and the following test are meant to check it's doing `using 1`

File: Mathlib/Tactic/Congr!.lean (hunk 2 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean]: Yes, and I rewrote this to try to clarify things.

File: Mathlib/Tactic/Congr!.lean (hunk 6 above)
Comment (kmill) [reply Mathlib/Tactic/Congr!.lean]: I just rewrote this whole docstring to give the plausibility algorithm

**Pull request closed without merging**